File: C++/breadth first search BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Perform BFS from starting node
void bfs(int start, const vector<vector<int>>& adj) {
    int n = adj.size();
    vector<bool> visited(n, false);
    queue<int> q;

    visited[start] = true;
    q.push(start);

    cout << "BFS traversal: ";
    while (!q.empty()) {
        int node = q.front();
        q.pop();
        cout << node << " ";

        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }
}

// ✅ Use-case
int main() {
    // Graph: 0-1-2-3-4
    vector<vector<int>> adj = {
        {1},        // 0
        {0, 2, 3},  // 1
        {1, 4},     // 2
        {1},        // 3
        {2}         // 4
    };
    bfs(0, adj);  // Output: BFS traversal: 0 1 2 3 4
}

File: Python/breadth-first search BFS.py
from collections import deque

def bfs(graph, start):
    """
    Breadth-First Search (BFS):
    - Visits all neighbors level by level (FIFO).
    - Uses a queue to explore nodes in breadth-first manner.
    - Time Complexity: O(V + E), where V = vertices, E = edges.
    """
    visited = set()  # To keep track of visited nodes
    queue = deque([start])  # Initialize queue with starting node
    result = []  # Store traversal order

    while queue:
        node = queue.popleft()  # Pop from front of queue
        if node not in visited:
            visited.add(node)  # Mark node as visited
            result.append(node)  # Add node to result
            for neighbor in graph.get(node, []):
                queue.append(neighbor)  # Add unvisited neighbors to queue

    return result

# ✅ Example graph and BFS call
graph = {
    'A': ['B', 'C'],
    'B': ['D', 'E'],
    'C': ['F'],
    'D': [], 'E': [], 'F': []
}
print("BFS Traversal:", bfs(graph, 'A'))  # Output: ['A', 'B', 'C', 'D', 'E', 'F']
